### PR TITLE
stm32: clock_configuration: Add test suite

### PR DESCRIPTION
**Pull request by erwango**
Add a test suite for stm32 clock configuration driver.

For now limited to stm32u5 as a demonstrator, it relies on STM32Cube HAL to check target clock configuration is programmed as expected. Overlays are added to propose various test configurations.

Additionally, add a change on stm32u5 VCO configuration.

### CODE REVIEW

File: drivers/clock_control/clock_stm32_ll_u5.c [hunk 1]
```diff
@@ -364,6 +377,10 @@ void config_src_sysclk_pll(LL_UTILS_ClkInitTypeDef s_ClkInitStruct)
 	LL_RCC_HSE_Disable();
 
 #elif STM32_PLL_SRC_HSI
+	set_regu_voltage(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
+	while (LL_PWR_IsActiveFlag_VOS() == 0) {
```
Comment (ABOSTM) [inline drivers/clock_control/clock_stm32_ll_u5.c:381]: Check is done in the function. can be removed.

File: drivers/clock_control/clock_stm32_ll_u5.c [hunk 2]
```diff
@@ -374,6 +391,10 @@ void config_src_sysclk_pll(LL_UTILS_ClkInitTypeDef s_ClkInitStruct)
 #elif STM32_PLL_SRC_HSE
 	int hse_bypass;
 
+	set_regu_voltage(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
+	while (LL_PWR_IsActiveFlag_VOS() == 0) {
+	}
```
Comment (ABOSTM) [inline drivers/clock_control/clock_stm32_ll_u5.c:396]: Dito

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_40.overlay [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+};
+
+&pll1 {
+	div-m = <4>;
+	mul-n = <10>;
+	div-q = <2>;
+	div-r = <1>;
+	clocks = <&clk_hsi>;
+	status = "okay";
+};
```
Comment (ABOSTM) [inline tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_40.overlay:26]: Don't you need to set:
```
&rcc {
	clock-frequency = <DT_FREQ_M(40)>;
```

File: drivers/clock_control/clock_stm32_ll_u5.c (hunk 1 above)
Comment (erwango) [reply drivers/clock_control/clock_stm32_ll_u5.c:381]: Sure, thanks!

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_160.overlay [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+};
+
+&pll1 {
+	div-m = <4>;
+	mul-n = <40>;
+	div-q = <2>;
+	div-r = <1>;
+	clocks = <&clk_hsi>;
+	status = "okay";
+};
```
Comment (ABOSTM) [inline tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_160.overlay:26]: Don't you need to set:
```
&rcc {
	clock-frequency = <DT_FREQ_M(160)>;
```

What about redefining systematically SysClock source and not relying on board default ?
```
&rcc {
	clocks = <&XXXX>;
```

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_40.overlay (hunk 3 above)
Comment (erwango) [reply tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_40.overlay:26]: Sure, I wonder how the test passed ...

Comment (erwango) [reply tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_40.overlay:26]: ok, got it: Test wasn't run.
Fixed

File: drivers/clock_control/clock_stm32_ll_u5.c (hunk 1 above)
Comment (erwango) [reply drivers/clock_control/clock_stm32_ll_u5.c]: Fixed

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_msis_24.overlay [hunk 5]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2021 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Warning: This overlay relies on initial nucleo_u575zi_q configuration
```
Comment (ABOSTM) [inline tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_msis_24.overlay:8]: > Warning: This overlay relies on initial **b_u585i_iot02a**  configuration

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_msis_24.overlay [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ *          For clarity, nodes are over written instead of deleted
+ *          Any change to board configuration has impact on this file
+ */
+
+&clk_lse {
+	status = "okay";
+};
```
Comment (ABOSTM) [inline tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_msis_24.overlay:15]: LSE is necessary for MSI autocalibration (PLLmode).
But I guess it is possible to work on MSI without PLL mode, so  without LSE.
Could be worth to have a new overlay with MSI but without LSE ?

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_msis_48.overlay [hunk 7]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (c) 2021 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Warning: This overlay relies on initial nucleo_u575zi_q configuration
```
Comment (ABOSTM) [inline tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_msis_48.overlay:8]: dito

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_160.overlay [hunk 8]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2021 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Warning: This overlay relies on initial nucleo_u575zi_q configuration
```
Comment (ABOSTM) [inline tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_160.overlay:8]: dito

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_msis_24.overlay (hunk 5 above)
Comment (erwango) [reply tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_msis_24.overlay]: Fixed

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_msis_24.overlay (hunk 6 above)
Comment (erwango) [reply tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_msis_24.overlay:15]: Sure, aim of current PR was to introduce the test suite, with the target of getting more tests in when possible.
I'm letting this proposal for a new PR if need is felt.

**Review by FRASTM — COMMENTED** [review]
else LGTM

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_160.overlay [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+};
+
+&clk_msis {					/* MSI disabled */
+	status = "disabled";
+	/delete-property/ msi-range;
+	/delete-property/ msi-pll-mode;
+};
+
+&clk_hsi {					/* HSI disabled */
+	status = "okay";
```
Comment (FRASTM) [inline tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_160.overlay:24]: should the hsi be **enabled** here ?

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_msi_80.overlay [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+	mul-n = <40>;
+	div-q = <2>;
+	div-r = <1>;
+	clocks = <&clk_msis>;
+	status = "okay";
+};
+
+&rcc {
+	clocks = <&pll1>;
+	clock-frequency = <DT_FREQ_M(80)>;	/* clck freq 40MHz */
```
Comment (FRASTM) [inline tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_msi_80.overlay:34]: is the clock freq 80MHz ?

**Review by ABOSTM — APPROVED** [review]
LGTM

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_160.overlay (hunk 9 above)
Comment (erwango) [reply tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_hsi_160.overlay:24]: Good catch, done

File: tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_msi_80.overlay (hunk 10 above)
Comment (erwango) [reply tests/drivers/clock_control/stm32_clock_configuration/boards/b_u585i_iot02a_pll_msi_80.overlay]: Good catch, done

**Review by FRASTM — APPROVED** [review]

**Pull request merged**